src/scvi/external/methylvi/_module.py

```diff
@@ -0,0 +1,269 @@
+            n_layers=n_layers,
+            n_hidden=n_hidden,
+            dropout_rate=dropout_rate,
+            return_dist=True,
+            var_activation=torch.nn.functional.softplus,  # Better numerical stability than exp
+        )
+
+        self.decoders = nn.ModuleDict()
+        for modality, num_features in zip(modalities, num_features_per_modality):
+            self.decoders[modality] = DecoderMETHYLVI(
+                n_latent,
+                num_features,
+                n_cat_list=cat_list,
+                n_layers=n_layers,
+                n_hidden=n_hidden,
+            )
+
+        if self.dispersion == "region":
+            self.px_gamma = torch.nn.ParameterDict(
+                {
+                    modality: nn.Parameter(torch.randn(num_features))
+                    for (modality, num_features) in zip(modalities, num_features_per_modality)
+                }
+            )
+
+    def _get_inference_input(self, tensors):
+        """Parse the dictionary to get appropriate args"""
+        mc = torch.cat(
+            [tensors[f"{x}_{METHYLVI_REGISTRY_KEYS.MC_KEY}"] for x in sorted(self.modalities)],
+            dim=1,
+        )
+        cov = torch.cat(
+            [tensors[f"{x}_{METHYLVI_REGISTRY_KEYS.COV_KEY}"] for x in sorted(self.modalities)],
+            dim=1,
+        )
+
+        batch_index = tensors[REGISTRY_KEYS.BATCH_KEY]
+
+        input_dict = {
+            METHYLVI_REGISTRY_KEYS.MC_KEY: mc,
+            METHYLVI_REGISTRY_KEYS.COV_KEY: cov,
+            "batch_index": batch_index,
+        }
+        return input_dict
+
+    def _get_generative_input(self, tensors, inference_outputs):
+        z = inference_outputs["z"]
+        batch_index = tensors[REGISTRY_KEYS.BATCH_KEY]
+
+        input_dict = {
+            "z": z,
+            "batch_index": batch_index,
+        }
+        return input_dict
+
+    @auto_move_data
+    def inference(self, mc, cov, batch_index, n_samples=1):
+        """
+        High level inference method.
+
+        Runs the inference (encoder) model.
+        """
+        # log the inputs to the variational distribution for numerical stability
+        mc_ = torch.log(1 + mc)
+        cov_ = torch.log(1 + cov)
+
+        # get variational parameters via the encoder networks
+        # we input both the methylated reads (mc) and coverage (cov)
+        total_input = torch.cat((mc_, cov_), dim=-1)
+        qz, z = self.z_encoder(total_input, batch_index)
+        if n_samples > 1:
+            z = qz.sample((n_samples,))
+
+        outputs = {"z": z, "qz": qz}
+        return outputs
+
+    @auto_move_data
+    def generative(self, z, batch_index):
+        """Runs the generative model."""
+        # form the parameters of the BetaBinomial likelihood
+        px_mu, px_gamma = {}, {}
+        for modality in self.modalities:
+            px_mu[modality], px_gamma[modality] = self.decoders[modality](
+                self.dispersion, z, batch_index
```

Comment:
Can you explain the modalities? And why they aren't fed to the encoder? Is it for different technologies? Are the cells with multiple modalities?